spmd/tensor/parallel/api.py

```diff
@@ -76,12 +76,69 @@ def _shard_self_attn_params(name: str, module: nn.Module) -> None:
                 module.register_module(n, tp_multi_head_attention)
 
 
+def _isMLP(module: nn.Module) -> bool:
+    # We assume that the structure of MLP to shard defined in object way as below:
+    # Linear -> ReLU -> Linear
+    submodules = list(module.children())
+    return (
+        len(submodules) == 3
+        and isinstance(submodules[0], nn.Linear)
+        and isinstance(submodules[1], nn.ReLU)
+        and isinstance(submodules[2], nn.Linear)
+    )
+
+
+def _gradient_hook(param: DTensor, grad: DTensor) -> None:
+    param._local_tensor.grad = grad._local_tensor
+
+
+def _shard_mlp(name: str, module: nn.Module, device_mesh: DeviceMesh) -> None:
+    if _isMLP(module):
+        col_wise_sharding: Sequence[Placement] = [Shard(0)]
+        row_wise_sharding: Sequence[Placement] = [Shard(1)]
+        replicate: Sequence[Placement] = [Replicate()] * device_mesh.ndim
+        submodules = list(module.children())
+
+        # shard Linear layer 1
+        m = submodules[0]
+        if isinstance(m, nn.Linear):
+            sharded_weight = nn.Parameter(
+                distribute_tensor(m.weight, device_mesh, col_wise_sharding)
+            )
+            sharded_bias = nn.Parameter(
+                distribute_tensor(m.bias, device_mesh, col_wise_sharding)
+            )
+            m.register_parameter("weight", sharded_weight)
+            m.register_parameter("bias", sharded_bias)
+            # note: this hook enables access to m.weight._local_tensor.grad
+            m.weight.register_hook(functools.partial(_gradient_hook, m.weight))
+
+        # shard Linear layer 2
+        m = submodules[2]
```

Comment:
It's hacky to assume that the 3rd submodule to be Linear. We can just iterate through all submodules and get its index.